// novus/lib/usermanager.js
var io = require('socket.io'),
	crypto = require('crypto'),
	mongo = require('mongoskin'),
	dbUsers = mongo.db('127.0.0.1:27017/exsto').collection('users');

exports.bind = function(socket) {
	socket.on('login', function(data) {
		authUser(data.name, data.password, socket);
	});
};

exports.getUser = function(conn) {
	var i = 0,
		ln = users.length;
	for(; i < ln; i++) {
		if(users[i].conn == conn) {
			return users[i];
		}
	}
};

exports.findUser = function(name) {
	var ret = null;
	
	users.forEach(function (user, index, array) {
		if(user.name == name) {
			ret = user;
			return false;
		}
	});
	
	return ret;
};

var users = [];

function authUser(name, password, conn) {
	password = crypto.createHash('md5').update(password).digest('hex');
	dbUsers.findOne({ name: name, password: password }, function (err, user) {
		if(user != null) {
			users.push(new User(name, conn));
			console.log('User joined: ' + name);
			
			conn.emit('login', { success: true });
		} else {
			conn.emit('login', { success: false });
		}
	});
}

function User(name, conn) {
	this.name = name;
	this.conn = conn;
	this.authenticated = true;
	this.currentRoom = -1;
};
